Add an easy way to register Villager and Wandering Trader trades

canitzp opened the pull request:
As said in #70 this is a very trivial implimentation for Trade registering, but it makes the whole thing a lot easier.
Also when the forge modloader is used, the corresponding events (VillagerTradesEvent & WandererTradesEvent) are used.

canitzp committed:
Add TradeRegistry to add Trades a bit more easily. Uses the VillagerTradesEvent on forge

canitzp committed:
Added TradeRegistry#registerTradeForWanderer which uses the WandererTraderEvent on forge

canitzp committed:
Added javadoc

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/architectury/architectury-api?pullRequest=84) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/architectury/architectury-api?pullRequest=84) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/architectury/architectury-api?pullRequest=84) it.</sub>

shedaniel commented on fabric/src/main/java/me/shedaniel/architectury/registry/fabric/TradeRegistryImpl.java:
```diff
@@ -0,0 +1,31 @@
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.math.NumberUtils;
+
+public class TradeRegistryImpl {
```
This needs to use the fabric trade registry, or else it will unregister on their own (Not our fault technically, their api is really really awful)

canitzp replied:
I didn't know that it existed, but I changed it

canitzp committed:
Use Fabric own Trade implementation and fixed Test Mod

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+    private TradeRegistry(){}
+    
+    /**
+     * Register a Trade (ItemListing) for a villager by it's profession and level.
```
This could use a {@link}

shedaniel replied:
Somehow my previous draft reviews slipped through 😅

canitzp replied:
already done

shedaniel commented on forge/src/main/java/me/shedaniel/architectury/registry/forge/TradeRegistryImpl.java:
```diff
@@ -0,0 +1,65 @@
+public class TradeRegistryImpl {
+    
+    private static final Map<VillagerProfession, Int2ObjectMap<List<VillagerTrades.ItemListing>>> TRADES_TO_ADD = new HashMap<>();
+    private static final Int2ObjectMap<List<VillagerTrades.ItemListing>> WANDERER_TRADES_TO_ADD = new Int2ObjectOpenHashMap<>();
```
This could use a Boolean2ObjectMap

canitzp replied:
already done

Juuxel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+    private TradeRegistry(){}
+    
+    /**
+     * Register a Trade (ItemListing) for a villager by it's profession and level.
```
```suggestion
     * Register a trade ({@link VillagerTrades.ItemListing}) for a villager by its profession and level.
```

Juuxel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+     * @param level The level the villager needs. Range is 1 to 5 with 1 being the lowest level and 5 the highest.
+     * @param trades The trades to add to this profession at the specified level.
+     *               
+     * When the Modloader is forge the {@code VillagerTradesEvent} event is used.
```
```suggestion
     * When the mod loader is Forge, the {@code VillagerTradesEvent} event is used.
```

Juuxel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    /**
+     * Register a Trade (ItemListing) to a wandering trader by its rarity.
```
```suggestion
     * Register a trade ({@link VillagerTrades.ItemListing}) to a wandering trader by its rarity.
```

Juuxel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+     * @param rare Is this a rare trade. Rare trades only have a five times lower chance of being used.
+     * @param trades The trades to add to the wandering trader.
+     *
+     * When the Modloader is forge the {@code WandererTradesEvent} event is used.
```
```suggestion
     * When the mod loader is Forge, the {@code WandererTradesEvent} event is used.
```

Juuxel commented on common/src/main/java/me/shedaniel/architectury/trade/SimpleTrade.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Random;
+
+public class SimpleTrade implements VillagerTrades.ItemListing {
```
IMO this class should be documented (including why it's needed)

canitzp replied:
Done

Juuxel commented on forge/src/main/java/me/shedaniel/architectury/registry/forge/TradeRegistryImpl.java:
```diff
@@ -0,0 +1,65 @@
+public class TradeRegistryImpl {
+    
+    private static final Map<VillagerProfession, Int2ObjectMap<List<VillagerTrades.ItemListing>>> TRADES_TO_ADD = new HashMap<>();
+    private static final Int2ObjectMap<List<VillagerTrades.ItemListing>> WANDERER_TRADES_TO_ADD = new Int2ObjectOpenHashMap<>();
```
`Boolean2ObjectMap` would make more sense here (as the key is pretty much the value of `rare`)

canitzp replied:
A Boolean2ObjectMap does not exist within fastutil

Juuxel replied:
Ah, I'd just use two list fields then. The outer map isn't really needed.

canitzp replied:
done

canitzp committed:
Update common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

canitzp committed:
Update common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

canitzp committed:
Update common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

canitzp committed:
Update common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

canitzp committed:
Added javadoc for SimpleTrade

canitzp committed:
Merge remote-tracking branch 'origin/villager_trades' into villager_trades

canitzp committed:
Use two lists instead of Int2ObjectMap

Juuxel commented on common/src/main/java/me/shedaniel/architectury/registry/TradeRegistry.java:
```diff
@@ -0,0 +1,38 @@
+     * When the mod loader is Forge, the {@code WandererTradesEvent} event is used.
+     */
+    @ExpectPlatform
+    public static void registerTradeForWanderer(boolean rare, VillagerTrades.ItemListing... trades){
```
I guess this should use `WanderingTrader` instead of `Wanderer` (I think that's a bad Forge name, actually?)

shedaniel replied:
Or try to be consistent with the method above?
registerWanderingTraderTrade, it may be a mouthful

canitzp replied:
done

Juuxel reviewed (approved):
Looks good apart from a minor naming thing.

shedaniel commented on forge/src/main/java/me/shedaniel/architectury/registry/forge/TradeRegistryImpl.java:
```diff
@@ -0,0 +1,65 @@
+    
+    public static void registerVillagerTrade(VillagerProfession profession, int level, VillagerTrades.ItemListing... trades) {
+        if(level < 1 || level > 5){
+            throw new RuntimeException("Villager Trade level has to be between 1 and 5!");
```
IllegalArgumentException, and print the current invalid value (Change this in the fabric one as well)

canitzp replied:
done

shedaniel reviewed (approved):
Other than some small nitpacks, this is alright, I'll fix the code style and licenses on merge

canitzp committed:
Use "registerTradeForWanderingTrader" instead of "registerTradeForWanderer" for better clarification

canitzp committed:
Use IllegalArgumentException instead of RuntimeException

shedaniel committed:
Remove level limit (Mods may be able to remove this restriction in VillagerData#canLevelUp), Clean up forge's implementation

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Clean up TestTrades and add licenses

Signed-off-by: shedaniel <daniel@shedaniel.me>

canitzp commented:
Maybe it is a idea to add the level check back, but only check if the passed level is greater or equal 1? I added this bounds check, because I added a level 0 in the test mod first and after half an hour wasted debugging, I realised that the lowest level is 1 and not 0 🤦 .
I understand that adding level is possible, so the upper bound is removed

shedaniel commented:
That sounds plausible, I will do that.

MaxNeedsSnacks committed:
[ciskip] Reintroduce lower bound validation for level

MaxNeedsSnacks committed:
Merge remote-tracking branch 'architectury/1.16' into villager_trades
